### PR TITLE
Spynode and Serialization Updates

### PR DESCRIPTION
**Pull request by KarlTheProgrammer**
Add merkle trees, merkle proofs, and encryption streams for the new spynode.
Update bitcoin object serialization.

### CODE REVIEW

File: bitcoin/signature.go [hunk 1]
```diff
@@ -173,6 +174,149 @@ func SignatureFromCompact(s string) (Signature, error) {
...
+	// sanity check length parsing
+	if index != len(b) {
+		return fmt.Errorf("Signature has bad final length %v != %v", index, len(b))
+	}
+
+	// Verify also checks this, but we can be more sure that we parsed
+	// correctly if we verify here too.
+	// FWIW the ecdsa spec states that R and S must be | 1, N - 1 |
+	// but crypto/ecdsa only checks for Sign != 0. Mirror that.
+	if s.R.Sign() != 1 {
```
Comment (scottjbarr) [inline bitcoin/signature.go:291]: This whole block could be reduced to a single "validation" func.

This would also keep SonarCloud happy.

Comment (KarlTheProgrammer) [reply bitcoin/signature.go:291]: Do you mean check the 4 mathematical validity tests in one if statement?

File: wire/merkle_proof.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import (
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/tokenized/pkg/bitcoin"
+
+	"github.com/pkg/errors"
+)
+
+type MerkleProof struct {
```
Comment (scottjbarr) [inline wire/merkle_proof.go:14]: Deserves a good comment.

File: wire/merkle_tree.go [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+				mp.AddHash(left, next)
+			}
+		}
+	}
+
+	// Append new layer
+	t.layers = append(t.layers, newMerkleNodeLayer(next))
+}
+
+func (t *MerkleTree) RootHash() bitcoin.Hash32 {
```
Comment (scottjbarr) [inline wire/merkle_tree.go:90]: Please add comment.

File: wire/merkle_proof.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+func NewMerkleProof(txid bitcoin.Hash32) *MerkleProof {
+	return &MerkleProof{
+		root:  txid,
+		TxID:  &txid,
+		Index: -1,
+		depth: 1,
+	}
+}
+
+func (p *MerkleProof) AddHash(hash, newRoot bitcoin.Hash32) {
```
Comment (scottjbarr) [inline wire/merkle_proof.go:41]: Please add comment to all public funcs in this file.

File: wire/merkle_tree.go [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+	}
+
+	if next == nil {
+		return bitcoin.Hash32{} // zero hash should never be valid
+	}
+	return *next
+}
+
+// FinalizeMerkleProofs finalizes the merkle proofs and returns the root hash and the merkle proofs.
+func (t MerkleTree) FinalizeMerkleProofs() (bitcoin.Hash32, []*MerkleProof) {
```
Comment (scottjbarr) [inline wire/merkle_tree.go:147]: SonarCloud gives this func 52 out 15 allowed.

The loop is long, and deep. Please consider ways to simplify.

File: wire/merkle_tree.go [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+		hashIndex := layer.count - len(layer.hashes) + 1
+		for _, hash := range layer.hashes {
+			fmt.Printf("%s%d : %s\n", indent, hashIndex, hash.String())
+			hashIndex++
+		}
+	}
+}
+
+// AddHash adds a new hash to the merkle tree.
+func (t *MerkleTree) AddHash(hash bitcoin.Hash32) {
```
Comment (scottjbarr) [inline wire/merkle_tree.go:47]: SonarCloud gives this 19 out of 15. Some fairly easy ways to reduce the complexity.

File: wire/merkle_tree.go (hunk 3 above)
Comment (KarlTheProgrammer) [reply wire/merkle_tree.go:90]: I am not sure why it wouldn't be obvious that it returns the root hash of the merkle tree, but I will add a comment.

File: wire/merkle_tree.go (hunk 5 above)
Comment (KarlTheProgrammer) [reply wire/merkle_tree.go:147]: I was able to move the hash calculation block of code out to another function with less than a 1% performance hit.

File: wire/merkle_tree.go (hunk 6 above)
Comment (KarlTheProgrammer) [reply wire/merkle_tree.go:47]: Moved this functions hash calculation block out too.

File: bitcoin/signature.go (hunk 1 above)
Comment (scottjbarr) [reply bitcoin/signature.go:291]: Moving that block of checks into a function would drop the overalll func complexity, and allow the "validation/guard" to be testable.

Comment (KarlTheProgrammer) [reply bitcoin/signature.go:291]: oh, I see. the code is duplicated too, so it is a good function.

**Pull request merged**